### PR TITLE
Updates for sorting rows in export

### PR DESCRIPTION
**Pull request by nickpell**
This implements a feature on the Solutions roadmap to improve the manageability of the Edit Export page.  All changes are behind a feature flag.  These changes can be tested [here](https://staging.commcarehq.org/a/staging/data/export/custom/new/case/edit/808b0fac8e4d46a5126cf3b04026c26f/).

Features included:

Drag and sort multiple rows at a time:
3n7qylealp

Select a range of rows using shift-click:
wqeytfbux8

Send a row to the top:
pfv9toahtv

Send a row to the bottom of selected rows:
viqsvj1she

@biyeun / @orangejenny Any UI feedback would be appreciated.
cc: @proteusvacuum

### CODE REVIEW

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (nickpell) [inline corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: I'd like to maintain the original order of selected rows, but currently the row that is clicked and dragged ends up first.

File: corehq/apps/hqwebapp/static/hqwebapp/css/export.css
Comment (nickpell) [inline corehq/apps/hqwebapp/static/hqwebapp/css/export.css]: Any suggestion on how to prevent this property from being overridden without having to use ```!important```?

**Comment by emord** [comment]
Have you tested this on a project with more case properties? And with a GS-level laptop?

I recall looking into something like this before and having it destroy performance past a certain amount of draggable elements, though it was probably a couple of years ago. might be good to have a baseline for when this gets slows the page down

**Comment by nickpell** [comment]
@emord I'm not surprised there are performance issues at some point, since the old and the new versions both contain some O(N) operations.  I'll get a baseline for when this page stops being usable.

**Comment by orangejenny** [comment]
@emord This page is already totally unusable past a certain number of case properties, we've gotten tickets from at least one SaaS project about it. Worth testing out on a large project, yeah, but unless it makes the "too many properties" number dramatically lower it's probably fine.

@nickpell Is this going through UAT? As mentioned, the usability of this page isn't just a solutions issue, and I think the usability needs of solutions projects and large saas projects are well aligned, so I'd like to not have this under a feature flag.

**Comment by orangejenny** [comment]
Haven't reviewed code yet, but feedback based on playing with it for a few minutes:
- The up and down arrows should be buttons, there isn't a click affordance
- Similarly, there isn't a drag affordance. The best thing would be to mimic the icon & cursor styling using elsewhere in HQ, like in module view's case detail property list.
- Different color for selected rows; the blue is a bit intense. In general, grab colors from https://www.commcarehq.org/styleguide/colors/ - I think any of the light shades could work, preferably the grey or blue. @biyeun might have a more specific opinion.
- Definite scope increase...but what do you think the LOE would be of adding an input with the numeric rank? Like the screenshot in [this ticket](https://manage.dimagi.com/default.asp?269343) or the fantasy draft screenshot in one of the responses to [this question](https://ux.stackexchange.com/questions/112415/best-way-to-reorder-long-list-of-items) which is a nice setup (drag icon + input box for rank). I think the current additions are worthwhile but suspect this page is still going to be painful for long lists (that stackexchange link has some decent discussion of ways to deal with this - basically it's a difficult ui to do well, and the more tools users have, the better).

File: corehq/apps/hqwebapp/static/hqwebapp/css/export.css
Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/css/export.css]: agree, is this necessary? didn't seem to be when testing locally.

Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/css/export.css]: oops, didn't see you commented on this already @nickpell

Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/css/export.css]: What causes it to be overridden? You can prevent it by making this selector more specific - like `#customize-export .table-striped tr.selected-for-sort` - but it's helpful to know what specific selector is causing it to be overriden (so you don't end up making a ridiculously specific, hard to read selector that wasn't necessary).

**Comment by orangejenny** [comment]
Also curious where design/requirements came from - what is it that indicates that being able to drag multiple adjacent rows will make this page more usable? Were there alternatives discussed? What's the default sort order of the properties?

**Comment by orangejenny** [comment]
btw, having a sample set up on staging was quite helpful, thank you for taking the time to do that.

File: corehq/apps/hqwebapp/static/hqwebapp/css/export.css
Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/css/export.css]: Thought about this some more. The easiest way to deal with this is probably to not do your own style, but instead add a bootstrap class like `.info` to the row when it's selected - check out the [contextual classes](http://bootstrapdocs.com/v3.0.3/docs/css/#tables-contextual-classes) docs.

File: corehq/apps/export/templates/export/partial/new_customize_export_templates.html
Comment (kaapstorm) [inline corehq/apps/export/templates/export/partial/new_customize_export_templates.html]: Could I advocate for using "[fa-angle-double-down](https://fontawesome.com/icons/angle-double-down?style=solid)" (and "[fa-angle-double-up](https://fontawesome.com/icons/angle-double-up?style=solid)") to suggest that this moves rows to the top and the bottom instead of just one row up and down? I know the column header makes this clear. But who reads words? :upside_down_face:

File: corehq/apps/hqwebapp/static/hqwebapp/css/export.css
Comment (biyeun) [reply corehq/apps/hqwebapp/static/hqwebapp/css/export.css]: yes, bootstrap has its own classes for styling / highlighting table rows that we should please use here.

**Comment by biyeun** [comment]
@orangejenny @nickpell yes, I definitely agree the blue selected is very jarring. Also this blue is not part of the color palette. It is not a best practice to be applying hex colors to the stylesheets ever, that's what the color variables are for (e.g. `@cc-brand-mid`) :-)

Bootstrap already has a style for highlighting table rows. please use that.

**Review by biyeun — CHANGES_REQUESTED** [review]
please use bootstrap styling for highlighting rows

http://bootstrapdocs.com/v3.0.3/docs/css/#tables-contextual-classes

File: corehq/apps/export/templates/export/partial/new_customize_export_templates.html
Comment (biyeun) [reply corehq/apps/export/templates/export/partial/new_customize_export_templates.html]: seconded

File: corehq/apps/hqwebapp/static/hqwebapp/css/export.css
Comment (nickpell) [reply corehq/apps/hqwebapp/static/hqwebapp/css/export.css]: removed in place of bootstrap class

File: corehq/apps/export/templates/export/partial/new_customize_export_templates.html
Comment (nickpell) [reply corehq/apps/export/templates/export/partial/new_customize_export_templates.html]: updated

**Comment by orangejenny** [comment]
@nickpell ha, I forgot what the current state looks like. Yeah, I'd just keep that.

**Comment by nickpell** [comment]
@orangejenny 

> Is this going through UAT? As mentioned, the usability of this page isn't just a solutions issue, and I think the usability needs of solutions projects and large saas projects are well aligned, so I'd like to not have this under a feature flag.

It will go through UAT, and then be advertised to GS users as a featured flagged enhancement.  It will also go through Dev and if it's low LOE to be released to Saas users, we'll do that.

> Also curious where design/requirements came from - what is it that indicates that being able to drag multiple adjacent rows will make this page more usable?

This [spec](https://docs.google.com/document/d/1fPNVesUhm1gsI2xvcedP4Yl2qSGNdgd4y7pjwKtvQ8w/edit#heading=h.q3wyha58n4s0) was written to reduce pain-points that GS flagged during interview with Nestle/Clayton.

> What's the default sort order of the properties?

Not sure exactly - it uses the same default order as the current exports do.

File: corehq/apps/export/templates/export/partial/new_customize_export_templates.html
Comment (nickpell) [inline corehq/apps/export/templates/export/partial/new_customize_export_templates.html]: @orangejenny What do you think about adding this styling to the row to suggest that the rows can be dragged?

Comment (orangejenny) [reply corehq/apps/export/templates/export/partial/new_customize_export_templates.html]: If you put the vertical arrows drag icon back and have a move cursor on that, I think that's enough to indicate draggability. I'd consider changing this style to a hand cursor, so the rows look clickable, since the selecting behavior is hard to discover (normally I'd suggest adding a checkbox, but since these rows already have checkboxes for a different purpose that doesn't work).

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (kaapstorm) [inline corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: Does this work? I would think that `list.splice` would update the value stored in the observable, but `list().splice` would return the value, and then update the returned value instead of the observable itself. I haven't tested this, so I'm not sure.

Line 154 uses the first construct (`list.splice`) to update the value. If they're equivalent, then I think this is fine.

**Review by biyeun — APPROVED** [review]

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (nickpell) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: Thanks for pointing that out.  I updated the send to top/bottom KO code, and that also allowed me to get rid of extra jQuery code.

I tried using the observable itself in the sortable code, but ran into some issues.  This section mixes jQuery/KO and they don't play nicely.  From testing, it seems that updating ```list()``` updates the underlying data without notifying the UI, whereas updating ```list``` notifies the UI.

Comment (nickpell) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: Planning on returning to this part at a later time

**Comment by nickpell** [comment]
Could I get one last review?  I made some changes that were committed before @biyeun's approval but pushed afterwards, so github misleading shows the commits as before the approval.

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (nickpell) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: @orangejenny Any recommendation for how to format this very long JS line?

Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: addClass and removeClass both accept multiple, space-separated classes, so you can do `.addClass("selected-for-sort success last-clicked")`, and similarly for the removeClass calls.

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: How about `multirow_sortable`? `multi_sortable` sounds like multisort, which I think is more commonly understood to mean sorting by multiple criteria.

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: So this comment gives me a clue to look up the source for makeTemplateValueAccessor, but I'd argue that instead of matching the code closely it'd be better to make this code slightly more readable - use _.isArray instead of `typeof unwrappedValue.length === "number"`, possibly copy over the comments from makeTemplateValueAccessor. There's a good chance this code will get out of "sync" with makeTemplateValueAccessor, might as well make it readable.

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: A little bit of actual commenting would help; this is a long bit of js to comprehend.

**Review by orangejenny — APPROVED** [review]
Left a few comments, but none of them are blocking, feel free to merge if you want to get this into UAT. I'm somewhat worried that the multi_sortable binding is going to be difficult to debug & maintain because it's long and mixes knockout with jquery, but I don't have any brilliant suggestions. Have you thought about writing tests for it? I think it'd be fairly straightforward to set up a test and the challenge will likely be getting drag & drop working in a test environment. A brief search turned up [this mock](https://github.com/andywer/drag-mock) which might be useful.

**Comment by nickpell** [comment]
Merging; will address remaining feedback in a separate PR

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (nickpell) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: https://github.com/dimagi/commcare-hq/pull/20895/commits/5114056217c5b8baf2f995e0edc978a8a0f1a1a4

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (nickpell) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: I think this handler might undergo significant changes before the end of the quarter, so I'm going to hold off on commenting for an audience who isn't me until it's more finalized.  If by the second week of July I haven't done so, I welcome you to hold me accountable :)

**Comment by nickpell** [comment]
@orangejenny Your point about maintainability and test coverage is well taken.  I'm going to bring this up at the Solutions meeting tomorrow (to see how this fits in with timelines) and I'll get back to you.

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (orangejenny) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: Makes sense, will do ;)

**Comment by orangejenny** [comment]
@nickpell Great, that sounds good. Thanks for the followup PR dealing with the smaller feedback.

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js
Comment (nickpell) [reply corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js]: https://github.com/dimagi/commcare-hq/pull/21299

**Pull request merged**
